File: day_2/puzzle.c
#include<stdio.h>
#include<stdlib.h>
#include<string.h>

int has_n_letters(char *line, int n) {
    int len = strlen(line);
    int chars[26] = {0};
    for (int i = 0; i < len; i++) {
        // Implicit conversion from character ASCII code to integer.
        chars[line[i] - 'a'] += 1;
    }
    // If any letter was seen multiple times, that array element is > 1.
    // Check if it is what we want. Return 1 if it is.
    for (int i = 0; i < 26; i++) {
        if(chars[i] == n) {
            return 1;
        }
    }
    return 0;
}

int has_one_diff(char* code_1, char* code_2, int print) {
    // The puzzle input specifies that codes are of the same length.
    // We just calculate one and proceed from there.
    int len = strlen(code_1);
    int diff = 0;
    for(int i = 0; i < len; i++) {
        if(code_1[i] != code_2[i]) {
            diff += 1;
        } else {
            // If we want to print the code, do it.
            if (print) {
                printf("%c", code_1[i]);
            }
        }
    }
    if (diff == 1) {
        return 1;
    }
    return 0;
}

int main() {

    FILE *fp;
    char input[256];

    int total_twos = 0;
    int total_threes = 0;

    int total_inputs = 0;
    char **input_strings = malloc(sizeof(char*));

    int code_found = 0;
    int checksum = 0;


    // Read the file.
    fp = fopen("./input.txt", "r");
    if (fp == NULL) {
        perror("input.txt");
        exit(EXIT_FAILURE);
    }

    // Store all values in a buffer.
    while((fscanf(fp, "%s\n", input)) != EOF) {
        total_twos += has_n_letters(input, 2);
        total_threes += has_n_letters(input, 3);

        total_inputs++;
        input_strings = realloc(input_strings, sizeof(char*) * total_inputs);
        input_strings[total_inputs-1] = malloc((strlen(input)+1) * sizeof(char));
        strcpy(input_strings[total_inputs-1], input);
    }
    fclose(fp);

    // Print the checksum.
    checksum = total_twos*total_threes;
    printf("The checksum is: %d\n", checksum);

    // Now find the strings that differ by just one.
    // This is again O(n*n) since we're checking each code against all others.
    for(int i = 0; i < total_inputs; i++) {
        for(int j = i; j < total_inputs; j++) {
            if(has_one_diff(input_strings[i], input_strings[j], 0)) {
                printf("The expected code is: ");
                has_one_diff(input_strings[i], input_strings[j], 1);
                printf("\n");
                code_found = 1;
                break;
            }
        }
        if (code_found) {
            break;
        }
    }

    // Free all the allocated memory.
    for(int i = 0; i < total_inputs; i++) {
        free(input_strings[i]);
    }
    free(input_strings);

    return 0;
}

File: day_2/puzzle.py
"""
AOC 2018, Day 2
"""

__author__ = "adibis"
__version__ = "0.1.0"
__license__ = "MIT"

def gen_characters(code):
    characters = dict()
    for letter in code:
        if letter in characters:
            characters[letter] += 1
        else:
            characters[letter] = 1

    return characters

def has_n_letters(code, n):
    characters = gen_characters(code)
    for key in characters:
        if characters[key] == n:
            return True
    return False

def get_code(code_1, code_2):
    diff = ""
    key = ""
    for a, b in zip(code_1, code_2):
        if a != b:
            diff += a+b
        else:
            key += a
    if len(diff) == 2:
        return key
    else:
        return False

def main():
    """Find the checksum of boxes and find two boxes which differ by 1"""
    with open ("input.txt", 'r') as f:
        read_data = f.readlines()

    two_letters = 0
    three_letters = 0
    checksum = 0
    key_found = False

    for code in read_data:
        if has_n_letters(code, 2):
            two_letters += 1
        if has_n_letters(code, 3):
            three_letters += 1

    checksum = two_letters * three_letters
    print("Final checksum: " + str(checksum))

    for code_1 in read_data:
        for code_2 in read_data:
            key = get_code(code_1, code_2)
            if key:
                print ("Matching Boxes: " + key )
                key_found = True
                break
        if key_found:
            break


if __name__ == "__main__":
    """ This is executed when run from the command line """
    main()
